modules/foreman_host.py

```diff
@@ -0,0 +1,217 @@
+    organization:
+        description:
+            - Organization
+        required: true
+    location:
+        description:
+            - Location
+        required: true
+    architecture
+        description:
+            - Architecture
+        required: true
+    operatingsystem
+        description:
+            - Operating System
+        required: true
+'''
+
+EXAMPLES = '''
+- name: "Create Host"
+  foreman_host:
+    username: "admin"
+    password: "admin"
+    server_url: "https://foreman.example.com"
+    hostname: "hostname"
+    mac: "00:00:00:00:00:00"
+    organization: "Default Organization"
+    location: "Default Location"
+    architecture: "x86_64"
+    operatingsystem: "Fedora 26"
+  delegate_to: localhost
+'''
+
+RETURN = '''# '''
+
+try:
+    from nailgun import entities, entity_fields, entity_mixins
+    from nailgun.config import ServerConfig
+    HAS_NAILGUN_PACKAGE = True
+except:
+    HAS_NAILGUN_PACKAGE = False
+
+
+class NailGun(object):
+
+    def __init__(self, server, entities, module):
+        self._server = server
+        self._entities = entities
+        self._module = module
+        entity_mixins.TASK_TIMEOUT = 1000
+
+
+    def create_host(self, hostname, mac, organization, location, architecture, operatingsystem):
+        location_id = self.find_location(location)
+        organization_id = self.find_organization(organization)
+        architecture_id = self.find_architecture(architecture)
+        operatingsystem_id = self.find_operatingsystem(operatingsystem)
+
+        host = self._entities.Host(self._server, name=hostname, organization=organization_id, location=location_id, mac=mac, architecture=architecture_id, operatingsystem=operatingsystem_id)
+        response = host.search(set(), {'search': 'name={}'.format(hostname)})
+        if len(response) == 0:
+            host.create()
+        else:
+            #host_id = response[0]
+            host_id = response[0]
+            host = self._entities.Host(self._server, name=hostname, id=host_id.id, organization=organization_id, location=location_id, mac=mac, architecture=architecture_id, operatingsystem=operatingsystem_id)
+            host.update()
+
+    def find_operatingsystem(self, name, **params):
+        os = self._entities.OperatingSystem(self._server, name=name, **params)
+        response = os.search(set(), {'search': 'name={}'.format(name)})
+        
+        if len(response) == 1:
+            return response[0]
+        else:
+            self._module.fail_json(msg="No operating found for %s" % name)
+
+    def find_architecture(self, name, **params):
+        org = self._entities.Architecture(self._server, name=name, **params)
+        response = org.search(set(), {'search': 'name={}'.format(name)})
+
+        if len(response) == 1:
+            return response[0]
+        else:
+            self._module.fail_json(msg="No architecture found for %s" % name)
+
+    def find_organization(self, name, **params):
```

Comment:
this already exists in https://github.com/theforeman/foreman-ansible-modules/blob/master/module_utils/ansible_nailgun_cement.py